README.md
# pytz

Brings the IANA tz database into Python. This library allows accurate and
cross platform timezone calculations.

pytz contains generated code, and this branch generates it. The actual
pytz code and documentation can be found in the src/ directory.

## Usage Information / Documentation

See [the pytz README](src/README.rst).

## pytz for Enterprise

Available as part of the Tidelift Subscription.

The maintainers of pytz and thousands of other packages are working with Tidelift to deliver commercial support and maintenance for the open source dependencies you use to build your applications. Save time, reduce risk, and improve code health, while paying the maintainers of the exact dependencies you use. [Learn more.](https://tidelift.com/subscription/pkg/pypi-pytz?utm_source=pypi-pytz&utm_medium=referral&utm_campaign=enterprise&utm_term=repo)


## Release process ##

    1) Untar upstream tarballs into elsie/src
    2) Update VERSION & OLSON_VERSION in src/pytz/__init__.py, and EXPECTED_VERSION in
       src/pytz/tests/test_tzinfo.py
    3) make test
    4) make dist



tz/NEWS
News for the tz database

Release 2021c - 2021-10-01 14:21:49 -0700

  Briefly:
    Revert most 2021b changes to 'backward'.
    Fix 'zic -b fat' bug in pre-1970 32-bit data.
    Fix two Link line typos.
    Distribute SECURITY file.

    This release is intended as a bugfix release, to fix compatibility
    problems and typos reported since 2021b was released.

  Changes to Link directives

    Revert almost all of 2021b's changes to the 'backward' file,
    by moving Link directives back to where they were in 2021a.
    Although 'zic' doesn't care which source file contains a Link
    directive, some downstream uses ran into trouble with the move.
    (Problem reported by Stephen Colebourne for Joda-Time.)

    Fix typo that linked Atlantic/Jan_Mayen to the wrong location
    (problem reported by Chris Walton).

    Fix 'backzone' typo that linked America/Virgin to the wrong
    location (problem reported by Michael Deckers).

  Changes to code

    Fix a bug in 'zic -b fat' that caused old timestamps to be
    mishandled in 32-bit-only readers (problem reported by Daniel
    Fischer).

  Changes to documentation

    Distribute the SECURITY file (problem reported by Andreas Radke).


Release 2021b - 2021-09-24 16:23:00 -0700

  Briefly:
    Jordan now starts DST on February's last Thursday.
    Samoa no longer observes DST.
    Merge more location-based Zones whose timestamps agree since 1970.
    Move some backward-compatibility links to 'backward'.
    Rename Pacific/Enderbury to Pacific/Kanton.
    Correct many pre-1993 transitions in Malawi, Portugal, etc.
    zic now creates each output file or link atomically.
    zic -L no longer omits the POSIX TZ string in its output.
    zic fixes for truncation and leap second table expiration.
    zic now follows POSIX for TZ strings using all-year DST.
    Fix some localtime crashes and bugs in obscure cases.
    zdump -v now outputs more-useful boundary cases.
    tzfile.5 better matches a draft successor to RFC 8536.
    A new file SECURITY.

    This release is prompted by recent announcements by Jordan and Samoa.
    It incorporates many other changes that had accumulated since 2021a.
    However, it omits most proposed changes that merged all Zones
    agreeing since 1970, as concerns were raised about doing too many of
    these changes at once.  It does keeps some of these changes in the
    interest of making tzdb more equitable one step at a time; see
    "Merge more location-based Zones" below.

  Changes to future timestamps

    Jordan now starts DST on February's last Thursday.
    (Thanks to Steffen Thorsen.)

    Samoa no longer observes DST.  (Thanks to Geoffrey D. Bennett.)

  Changes to zone name

    Rename Pacific/Enderbury to Pacific/Kanton.  When we added
    Enderbury in 1993, we did not know that it is uninhabited and that
    Kanton (population two dozen) is the only inhabited location in
    that timezone.  The old name is now a backward-compatility link.

  Changes to past timestamps

    Correct many pre-1993 transitions, fixing entries originally
    derived from Shanks, Whitman, and Mundell.  The fixes include:
      - Barbados: standard time was introduced in 1911, not 1932; and
	DST was observed in 1942-1944
      - Cook Islands: In 1899 they switched from east to west of GMT,
	celebrating Christmas for two days.  They (and Niue) switched
	to standard time in 1952, not 1901.
      - Guyana: corrected LMT for Georgetown; the introduction of
	standard time in 1911, not 1915; and corrections to 1975 and
	1992 transitions
      - Kanton: uninhabited before 1937-08-31
      - Niue: only observed -11:20 from 1952 through 1964, then went to
        -11 instead of -11:30
      - Portugal: DST was observed in 1950
      - Tonga: corrected LMT; the introduction of standard time in 1945,
        not 1901; and corrections to the transition from +12:20 to +13
        in 1961, not 1941
    Additional fixes to entries in the 'backzone' file include:
      - Enderbury: inhabited only 1860/1885 and 1938-03-06/1942-02-09
      - The Gambia: 1933 and 1942 transitions
      - Malawi: several 1911 through 1925 transitions
      - Sierra Leone: several 1913 through 1941 transitions, and DST
	was NOT observed in 1957 through 1962
    (Thanks to P Chan, Michael Deckers, Alexander Krivenyshev and
    Alois Treindl.)

    Merge more location-based Zones whose timestamps agree since 1970,
    as pre-1970 timestamps are out of scope.  This is part of a
    process that has been ongoing since 2013.  This does not affect
    post-1970 timestamps, and timezone historians who build with 'make
    PACKRATDATA=backzone' should see no changes to pre-1970 timestamps.
    When merging, keep the most-populous location's data, and move
    data for other locations to 'backzone' with a backward
    link in 'backward'.  For example, move America/Creston data to
    'backzone' with a link in 'backward' from America/Phoenix because
    the two timezones' timestamps agree since 1970; this change
    affects some pre-1968 timestamps in America/Creston because
    Creston and Phoenix disagreed before 1968.  The affected Zones
    are Africa/Accra, America/Atikokan, America/Blanc-Sablon,
    America/Creston, America/Curacao, America/Nassau,
    America/Port_of_Spain, Antarctica/DumontDUrville, and
    Antarctica/Syowa.

  Changes to maintenance procedure

    The new file SECURITY covers how to report security-related bugs.

    Several backward-compatibility links have been moved to the
    'backward' file.  These links, which range from Africa/Addis_Ababa
    to Pacific/Saipan, are only for compatibility with now-obsolete
    guidelines suggesting an entry for every ISO 3166 code.
    The intercontinental convenience links Asia/Istanbul and
    Europe/Nicosia have also been moved to 'backward'.

  Changes to code

    zic now creates each output file or link atomically,
    possibly by creating a temporary file and then renaming it.

src/pytz/__init__.py
'''
datetime.tzinfo timezone definitions generated from the
Olson timezone database:

    ftp://elsie.nci.nih.gov/pub/tz*.tar.gz

See the datetime section of the Python Library Reference for information
on how to use these modules.
'''

import collections
import datetime
import os.path
import re
import sys

from pytz.exceptions import AmbiguousTimeError
from pytz.exceptions import InvalidTimeError
from pytz.exceptions import NonExistentTimeError
from pytz.exceptions import UnknownTimeZoneError
from pytz.lazy import LazyDict, LazyList, LazySet  # noqa
from pytz.tzinfo import unpickler, BaseTzInfo
from pytz.tzfile import build_tzinfo


# The IANA (nee Olson) database is updated several times a year.
OLSON_VERSION = '2021c'
VERSION = '2021.3'  # pip compatible version number.
__version__ = VERSION

OLSEN_VERSION = OLSON_VERSION  # Old releases had this misspelling

__all__ = [
    'timezone', 'utc', 'country_timezones', 'country_names',
    'AmbiguousTimeError', 'InvalidTimeError',
    'NonExistentTimeError', 'UnknownTimeZoneError',
    'all_timezones', 'all_timezones_set',
    'common_timezones', 'common_timezones_set',
    'BaseTzInfo', 'FixedOffset',
]


if sys.version_info[0] > 2:  # Python 3.x

    # Python 3.x doesn't have unicode(), making writing code
    # for Python 2.3 and Python 3.x a pain.
    unicode = str

    def ascii(s):
        r"""
        >>> ascii('Hello')
        'Hello'
        >>> ascii('\N{TRADE MARK SIGN}') #doctest: +IGNORE_EXCEPTION_DETAIL
        Traceback (most recent call last):
            ...
        UnicodeEncodeError: ...
        """
        if type(s) == bytes:
            s = s.decode('ASCII')
        else:
            s.encode('ASCII')  # Raise an exception if not ASCII
        return s  # But the string - not a byte string.

else:  # Python 2.x

    def ascii(s):
        r"""
        >>> ascii('Hello')
        'Hello'
        >>> ascii(u'Hello')
        'Hello'
        >>> ascii(u'\N{TRADE MARK SIGN}') #doctest: +IGNORE_EXCEPTION_DETAIL
        Traceback (most recent call last):
            ...
        UnicodeEncodeError: ...
        """
        return s.encode('ASCII')


def open_resource(name):
    """Open a resource from the zoneinfo subdir for reading.

    Uses the pkg_resources module if available and no standard file
    found at the calculated location.

    It is possible to specify different location for zoneinfo
    subdir by using the PYTZ_TZDATADIR environment variable.
    """
    name_parts = name.lstrip('/').split('/')
    for part in name_parts:
        if part == os.path.pardir or os.path.sep in part:
            raise ValueError('Bad path segment: %r' % part)
    zoneinfo_dir = os.environ.get('PYTZ_TZDATADIR', None)
    if zoneinfo_dir is not None:
        filename = os.path.join(zoneinfo_dir, *name_parts)
    else:
        filename = os.path.join(os.path.dirname(__file__),
                                'zoneinfo', *name_parts)
        if not os.path.exists(filename):
            # http://bugs.launchpad.net/bugs/383171 - we avoid using this
            # unless absolutely necessary to help when a broken version of
            # pkg_resources is installed.
            try:
                from pkg_resources import resource_stream
            except ImportError:
                resource_stream = None

            if resource_stream is not None:
                return resource_stream(__name__, 'zoneinfo/' + name)
    return open(filename, 'rb')


def resource_exists(name):
    """Return true if the given resource exists"""
    try:
        if os.environ.get('PYTZ_SKIPEXISTSCHECK', ''):
            # In "standard" distributions, we can assume that
            # all the listed timezones are present. As an
            # import-speed optimization, you can set the
            # PYTZ_SKIPEXISTSCHECK flag to skip checking
            # for the presence of the resource file on disk.
            return True
        open_resource(name).close()
        return True
    except IOError:
        return False


_tzinfo_cache = {}


def timezone(zone):
    r''' Return a datetime.tzinfo implementation for the given timezone

    >>> from datetime import datetime, timedelta
    >>> utc = timezone('UTC')
    >>> eastern = timezone('US/Eastern')
    >>> eastern.zone
    'US/Eastern'
    >>> timezone(unicode('US/Eastern')) is eastern
    True
    >>> utc_dt = datetime(2002, 10, 27, 6, 0, 0, tzinfo=utc)
    >>> loc_dt = utc_dt.astimezone(eastern)
    >>> fmt = '%Y-%m-%d %H:%M:%S %Z (%z)'
    >>> loc_dt.strftime(fmt)
    '2002-10-27 01:00:00 EST (-0500)'
    >>> (loc_dt - timedelta(minutes=10)).strftime(fmt)
    '2002-10-27 00:50:00 EST (-0500)'
    >>> eastern.normalize(loc_dt - timedelta(minutes=10)).strftime(fmt)
    '2002-10-27 01:50:00 EDT (-0400)'
    >>> (loc_dt + timedelta(minutes=10)).strftime(fmt)
    '2002-10-27 01:10:00 EST (-0500)'

    Raises UnknownTimeZoneError if passed an unknown zone.

    >>> try:
    ...     timezone('Asia/Shangri-La')
    ... except UnknownTimeZoneError:
    ...     print('Unknown')
    Unknown

    >>> try:
    ...     timezone(unicode('\N{TRADE MARK SIGN}'))
    ... except UnknownTimeZoneError:
    ...     print('Unknown')
    Unknown

    '''
    if zone is None:
        raise UnknownTimeZoneError(None)

    if zone.upper() == 'UTC':
        return utc

    try:
        zone = ascii(zone)
    except UnicodeEncodeError:
        # All valid timezones are ASCII
        raise UnknownTimeZoneError(zone)

    zone = _case_insensitive_zone_lookup(_unmunge_zone(zone))
    if zone not in _tzinfo_cache:
        if zone in all_timezones_set:  # noqa
            fp = open_resource(zone)
            try:
                _tzinfo_cache[zone] = build_tzinfo(zone, fp)
            finally:
                fp.close()
        else:
            raise UnknownTimeZoneError(zone)

    return _tzinfo_cache[zone]


def _unmunge_zone(zone):
    """Undo the time zone name munging done by older versions of pytz."""
    return zone.replace('_plus_', '+').replace('_minus_', '-')


_all_timezones_lower_to_standard = None


def _case_insensitive_zone_lookup(zone):
    """case-insensitively matching timezone, else return zone unchanged"""
    global _all_timezones_lower_to_standard
    if _all_timezones_lower_to_standard is None:
        _all_timezones_lower_to_standard = dict((tz.lower(), tz) for tz in all_timezones)  # noqa
    return _all_timezones_lower_to_standard.get(zone.lower()) or zone  # noqa


ZERO = datetime.timedelta(0)
HOUR = datetime.timedelta(hours=1)


class UTC(BaseTzInfo):
    """UTC

    Optimized UTC implementation. It unpickles using the single module global
    instance defined beneath this class declaration.
    """
    zone = "UTC"

    _utcoffset = ZERO
    _dst = ZERO
    _tzname = zone

    def fromutc(self, dt):
        if dt.tzinfo is None:
            return self.localize(dt)
        return super(utc.__class__, self).fromutc(dt)

    def utcoffset(self, dt):
        return ZERO

    def tzname(self, dt):
        return "UTC"

    def dst(self, dt):
        return ZERO

    def __reduce__(self):
        return _UTC, ()

    def localize(self, dt, is_dst=False):
        '''Convert naive time to local time'''
        if dt.tzinfo is not None:
            raise ValueError('Not naive datetime (tzinfo is already set)')
        return dt.replace(tzinfo=self)

    def normalize(self, dt, is_dst=False):
        '''Correct the timezone information on the given datetime'''
        if dt.tzinfo is self:
            return dt
        if dt.tzinfo is None:
            raise ValueError('Naive time - no tzinfo set')
        return dt.astimezone(self)

    def __repr__(self):
        return "<UTC>"

    def __str__(self):
        return "UTC"


UTC = utc = UTC()  # UTC is a singleton


def _UTC():
    """Factory function for utc unpickling.

    Makes sure that unpickling a utc instance always returns the same
    module global.

    These examples belong in the UTC class above, but it is obscured; or in
    the README.rst, but we are not depending on Python 2.4 so integrating
    the README.rst examples with the unit tests is not trivial.

    >>> import datetime, pickle
    >>> dt = datetime.datetime(2005, 3, 1, 14, 13, 21, tzinfo=utc)
    >>> naive = dt.replace(tzinfo=None)
    >>> p = pickle.dumps(dt, 1)
    >>> naive_p = pickle.dumps(naive, 1)
    >>> len(p) - len(naive_p)
    17
    >>> new = pickle.loads(p)
    >>> new == dt
    True
    >>> new is dt
    False
    >>> new.tzinfo is dt.tzinfo
    True
    >>> utc is UTC is timezone('UTC')
    True
    >>> utc is timezone('GMT')
    False
    """
    return utc


_UTC.__safe_for_unpickling__ = True


def _p(*args):
    """Factory function for unpickling pytz tzinfo instances.

    Just a wrapper around tzinfo.unpickler to save a few bytes in each pickle
    by shortening the path.
    """
    return unpickler(*args)


_p.__safe_for_unpickling__ = True


def _zone_tab():
    """Load zone.tab and return each line in parsed form.

    >>> zone_tab[0] == ('AD', (42.5, 1.5166666666666666), 'Europe/Andorra', '')
    True
    >>> sys.version_info[0] == 2 and sys.version_info[1] < 6 or zone_tab[0].code == 'AD'
    True
    >>> [row for row in zone_tab if row[2] == 'America/Los_Angeles'][0][3] == 'Pacific'
    True
    """
    try:
        zonetype = collections.namedtuple('Zone', ('code', 'coordinates', 'zone', 'comment'))
    except AttributeError:
        zonetype = lambda *args: args
    lines = []
    try:
        fobj = open_resource('zone.tab')
        lines = fobj.readlines()
    finally:
        fobj.close()
    for line in lines:
        line = line.decode('UTF-8')
        if line.startswith('#'):
            continue
        pieces = line.rstrip().split(None, 3)
        code, coordinates, zone = pieces[:3]
        comment = len(pieces) == 4 and pieces[3] or ''
        if zone not in all_timezones_set:
            continue

        coordinates = _parse_iso6709(coordinates)
        if not coordinates:
            continue
        lat, lon = coordinates

        yield zonetype(code, (lat, lon), zone, comment)


zone_tab = LazyList(_zone_tab())


def _parse_iso6709(iso6709):
    match = re.match('^([-+][0-9]+)([-+][0-9]+)$', iso6709)
    if not match:
        return
    return map(_parse_iso6709_coord, match.groups())


def _parse_iso6709_coord(coord):
    if len(coord) % 2:
        deg = float(coord[:3])
        coord = coord[3:]
    else:
        deg = float(coord[:4])
        coord = coord[4:]
    div = 60
    while coord:
        deg += float(coord[:2]) / div
        div *= 60
        coord = coord[2:]
    return deg


class _CountryTimezoneDict(LazyDict):
    """Map ISO 3166 country code to a list of timezone names commonly used
    in that country.

    iso3166_code is the two letter code used to identify the country.

    >>> def print_list(list_of_strings):
    ...     'We use a helper so doctests work under Python 2.3 -> 3.x'
    ...     for s in list_of_strings:
    ...         print(s)

    >>> print_list(country_timezones['nz'])
    Pacific/Auckland
    Pacific/Chatham
    >>> print_list(country_timezones['ch'])
    Europe/Zurich
    >>> print_list(country_timezones['CH'])
    Europe/Zurich
    >>> print_list(country_timezones[unicode('ch')])
    Europe/Zurich
    >>> print_list(country_timezones['XXX'])
    Traceback (most recent call last):
    ...
    KeyError: 'XXX'

    Previously, this information was exposed as a function rather than a
    dictionary. This is still supported::

    >>> print_list(country_timezones('nz'))
    Pacific/Auckland
    Pacific/Chatham
    """
    def __call__(self, iso3166_code):
        """Backwards compatibility."""
        return self[iso3166_code]

    def _fill(self):
        data = {}
        for code, _, zone, _ in zone_tab:
            try:
                data[code].append(zone)
            except KeyError:
                data[code] = [zone]
        self.data = data


country_timezones = _CountryTimezoneDict()


class _CountryNameDict(LazyDict):
    '''Dictionary proving ISO3166 code -> English name.

    >>> print(country_names['au'])
    Australia
    '''
    def _fill(self):
        data = {}
        zone_tab = open_resource('iso3166.tab')
        try:
            for line in zone_tab.readlines():
                line = line.decode('UTF-8')
                if line.startswith('#'):
                    continue
                code, name = line.split(None, 1)
                data[code] = name.strip()
            self.data = data
        finally:
            zone_tab.close()


country_names = _CountryNameDict()


# Time-zone info based solely on fixed offsets

class _FixedOffset(datetime.tzinfo):

    zone = None  # to match the standard pytz API

    def __init__(self, minutes):
        if abs(minutes) >= 1440:
            raise ValueError("absolute offset is too large", minutes)
        self._minutes = minutes
        self._offset = datetime.timedelta(minutes=minutes)

    def utcoffset(self, dt):
        return self._offset

    def __reduce__(self):
        return FixedOffset, (self._minutes, )

    def dst(self, dt):
        return ZERO

    def tzname(self, dt):
        return None

    def __repr__(self):
        return 'pytz.FixedOffset(%d)' % self._minutes

    def localize(self, dt, is_dst=False):
        '''Convert naive time to local time'''
        if dt.tzinfo is not None:
            raise ValueError('Not naive datetime (tzinfo is already set)')
        return dt.replace(tzinfo=self)

    def normalize(self, dt, is_dst=False):
        '''Correct the timezone information on the given datetime'''
        if dt.tzinfo is self:
            return dt
        if dt.tzinfo is None:
            raise ValueError('Naive time - no tzinfo set')
        return dt.astimezone(self)


def FixedOffset(offset, _tzinfos={}):
    """return a fixed-offset timezone based off a number of minutes.

        >>> one = FixedOffset(-330)
        >>> one
        pytz.FixedOffset(-330)
        >>> str(one.utcoffset(datetime.datetime.now()))
        '-1 day, 18:30:00'
        >>> str(one.dst(datetime.datetime.now()))
        '0:00:00'

        >>> two = FixedOffset(1380)
        >>> two
        pytz.FixedOffset(1380)
        >>> str(two.utcoffset(datetime.datetime.now()))
        '23:00:00'
        >>> str(two.dst(datetime.datetime.now()))
        '0:00:00'

    The datetime.timedelta must be between the range of -1 and 1 day,
    non-inclusive.

        >>> FixedOffset(1440)
        Traceback (most recent call last):
        ...
        ValueError: ('absolute offset is too large', 1440)

        >>> FixedOffset(-1440)
        Traceback (most recent call last):
        ...
        ValueError: ('absolute offset is too large', -1440)

    An offset of 0 is special-cased to return UTC.

        >>> FixedOffset(0) is UTC
        True

    There should always be only one instance of a FixedOffset per timedelta.
    This should be true for multiple creation calls.

        >>> FixedOffset(-330) is one
        True
        >>> FixedOffset(1380) is two
        True

    It should also be true for pickling.

        >>> import pickle
        >>> pickle.loads(pickle.dumps(one)) is one
        True
        >>> pickle.loads(pickle.dumps(two)) is two
        True
    """
    if offset == 0:
        return UTC

    info = _tzinfos.get(offset)
    if info is None:
        # We haven't seen this one before. we need to save it.

        # Use setdefault to avoid a race condition and make sure we have
        # only one
        info = _tzinfos.setdefault(offset, _FixedOffset(offset))

    return info


FixedOffset.__safe_for_unpickling__ = True


def _test():
    import doctest
    sys.path.insert(0, os.pardir)
    import pytz
    return doctest.testmod(pytz)


if __name__ == '__main__':
    _test()
